### ISSUE TITLE
Allow doneActions to be invoked by name

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#2601]
Instead of

    Env.adsr(0.01, 0.03, 0.7, 0.3).kr(2)

I think it would improve readability to replace the number 2 with an English word:

    Env.adsr(0.01, 0.03, 0.7, 0.3).kr(\free)

Since it's too laborious to do this for every doneAction, I'd be alright with only implementing this sugar for the most common places where people use doneActions: Env:ar, Env:kr, EnvGen.ar, EnvGen.kr, and Done.kr.

**Comment by Sciss** [comment]
I don't understand - doesn't this already work automatically? That argument is just forwarded to `EnvGen`, and AFAIK it's fine to use the symbols for done-action.

**Comment by Sciss** [comment]
Ooops, I was mistaken

```
play { 
	Line.kr(0, 1, 10, \freeSelf);
	nil
}
```

I thought that worked (because it does in ScalaCollider). Bad memory. Would make sense indeed to apply something like `doneAction.asDoneActionConstant` (or similar).

**Comment by brianlheim** [comment]
What about this alternative:
```
Env.adsr(0.01, 0.03, 0.7, 0.3).kr(Done.freeSelf)
```

Benefits:
* Centralized
* Follows the enum-style pattern of `Char.nl` etc.
* Doesn't require creation of Symbol object
* No opportunity for typos to ruin your day

Cons:
* Bulkier than most SynthDef notations
* Adds to selector table

A similar idea would be to create an enum-ish class `DoneAction`. Basically I'm wondering if there's another option that requires neither UGens nor Integer/Number to be aware of this shortcut.

**Comment by vivid-synth** [comment]
> No opportunity for typos to ruin your day

I like the style, but how does it help with typos?

**Comment by brianlheim** [comment]
@vivid-synth I've had tons of silent bugs in the past caused by misspelling symbols or global variable names. In my experience, it's much less painful to debug a "does not understand" message rather than sift through code only to find that the reason your Synth isn't behaving as expected is that you've spelled `\gste` or `\veryLongDescripter` or `~environmentVariabel` wrong.

**Comment by snappizz** [comment]
It is easy to add a check to ensure that the symbol is valid.

**Comment by brianlheim** [comment]
Hmm. If that check lives inside every individual UGen with a doneAction argument, that's a lot of duplicated code. If the aliases belong to a separate object, none of the UGen code needs to change and the aliases can be used anywhere, even outside of UGen arguments.

**Comment by jamshark70** [comment]
> If that check lives inside every individual UGen with a doneAction argument, that's a lot of duplicated code.

Not if it's implemented as a method on Symbol, say `asDoneAction` -- this method could throw an error if it's invalid.

Then you'd also need the same method for Nil and SimpleNumber (and this would add the benefit that the number could be validated, which we're not doing now). I guess we'd need UGen as well, though I think it's pretty rare to modulate the doneAction input.

Given all those cases, polymorphism is the only reasonable way to do this.

**Comment by brianlheim** [comment]
I'm having a hard time seeing the need for polymorphism here; I thought we were just discussing a simple conversion from names to a small range of integers. If you wouldn't mind giving some use cases, that would help a lot. Also, this might be precedent in SC, but there's definitely more than one reasonable way to do this, and I think the benefits to the alternative I've outlined are still worthy of consideration. Polymorphism taken to an extreme can make the codebase bloated, so I'm just trying to play devil's advocate. :)

**Comment by jamshark70** [comment]
I was responding to the idea about using symbols directly (without a Done class) and validating the symbols. If we want to write, say, `EnvGen.kr(..., doneAction: \freeSelf)`, then .kr needs to convert the symbol by some bit of code, AND that bit of code would have to handle symbols, numbers, UGens and nil at least... and when you need a bunch of different classes to respond differently, but sensibly, to the same bit of code, polymorphism is the obvious choice.

**Comment by brianlheim** [comment]
Ok! Could you please give more details about how you would implement that? I can imagine a few different ways.

**Comment by jamshark70** [comment]
OK, I've just put a little sketch here: https://github.com/jamshark70/supercollider/commit/27e0d794366e3f8c0eda6e7ad8fb95ec405ae00c

A particular problem is summarizing some rather complex done-actions as short symbols, e.g. "11 -- free this synth and if the preceding node is a group then do g_deepFree on it, else free it." To be quite honest, when I read the documentation for such done actions, I think this whole feature request might be barking up the wrong tree. A symbolic name for this is very likely to be either misleading, or so long as to be inconvenient to type (and, as a symbol, it's not eligible for IDE autocompletion).

Take the editorial comment or leave it, but whatever the decision is, just make sure that it doesn't make matters worse. ("Making it worse" would include adding code that has to be maintained, but benefit is dubious.)

Test case:

```
// First block should throw an error on the last def
(
SynthDef(\testNil, {
	Out.kr(0, EnvGen.kr(Env.adsr, doneAction: nil))
}).add;

SynthDef(\testNumber, {
	Out.kr(0, EnvGen.kr(Env.adsr, doneAction: 2))
}).add;

SynthDef(\testSymbol, {
	Out.kr(0, EnvGen.kr(Env.adsr, doneAction: \freeSelf))
}).add;

SynthDef(\testUGen, { |doneAction = 2|
	Out.kr(0, EnvGen.kr(Env.adsr, doneAction: doneAction))
}).add;

SynthDef(\testWrongSymbol, { |doneAction = 2|
	Out.kr(0, EnvGen.kr(Env.adsr, doneAction: \fakeOne))
}).add;
)

// Should print [ 0, 2, 2, an OutputProxy ]
(
#[testNil, testNumber, testSymbol, testUGen].collect { |name|
	var env = SynthDescLib.global[name].def.children.detect(_.isKindOf(EnvGen));
	if(env.notNil) { env.inputs[4] } { "no env" }
};
)
```

This commit does not go through all the methods that have a doneAction argument. Here's a code block that finds all of the methods that would have to be updated (changing `doneAction` to `doneAction.asDoneAction`).

```
(
var result = List.new,
traverse = { |class|
	var method = class.methods.select { |m|
		if(m.argNames.isSequenceableCollection) {
			m.argNames.includes(\doneAction)
		} {
			false
		}
	};
	if(method.size > 0) {
		result.addAll(method);
	};
	class.subclasses.do { |cl| traverse.(cl) };
};
traverse.(Meta_UGen);
result.do(_.postln); ""
)

->
Meta_Line:ar
Meta_Line:kr
Meta_EnvGen:ar
Meta_EnvGen:kr
Meta_EnvGen:new1
Meta_DemandEnvGen:kr
Meta_DemandEnvGen:ar
Meta_LFGauss:ar
Meta_LFGauss:kr
Meta_Duty:ar
Meta_Duty:kr
Meta_TDuty:ar
Meta_TDuty:kr
Meta_PlayBuf:ar
Meta_PlayBuf:kr
Meta_Linen:kr
Meta_DetectSilence:ar
Meta_DetectSilence:kr
Meta_RecordBuf:ar
Meta_RecordBuf:kr
Meta_XLine:ar
Meta_XLine:kr
```

**Comment by Sciss** [comment]
Perhaps take some inspirations from the naming conventions in ScalaCollider: https://sciss.github.io/ScalaCollider/latest/api/#de.sciss.synth.DoneAction:

-    0 - doNothing
-    1 - pauseSelf
-    2 - freeSelf
-    3 - freeSelfPred
-    4 - freeSelfSucc
-    5 - freeSelfPredAll
-    6 - freeSelfSuccAll
-    7 - freeSelfToHead
-    8 - freeSelfToTail
-    9 - freeSelfPausePred
-    10 - freeSelfPauseSucc
-    11 - freeSelfPredDeep
-    12 - freeSelfSuccDeep
-    13 - freeAllInGroup
-    14 - freeGroup

**Comment by jamshark70** [comment]
(I'll admit to a little devil's-advocacy in my last remark. I just think it's worth being careful that this is not one of those cases where we think, "Oh yeah, great idea, we should have symbols for that!" and then run into problems with the details.)

I have no objection to changing my symbolic names. I'm often tone-deaf about that.

**Comment by brianlheim** [comment]
Thank you for this!

I guess my main worry is that adding that method call individually to each UGen that has a doneAction argument is duplication that has the potential to become out-of-date and obscure. Every time a new UGen with a doneAction is written, only the people who know about this trick will be able to add it in order to use what appears to be a codebase-consistent syntax hack.

Just to be totally clear, here's the implementation of what I'm suggesting:

```
DoneAction {
  const <doNothing = 0;
  const <pauseSelf = 1;
  const <freeSelf = 2;
  const <freeSelfPrev = 3;
...
  const <freeGroup = 14;
}
```

So it's quite compact, and fully encapsulates the concept without entangling other behavior. The IDE autocompletion remark is noteworthy, and solved really well by this approach. If you type in "DoneAction.", you'll immediately see a list of all the possible doneAction names.

As far as the need for this feature, I think the problem we're attempting to solve is twofold: (1) make code more readable—i.e. the text "freeSelf" vs the number "2"—and (2) reduce or eliminate trips through the documentation to look up the proper (or possible) done action names and codes. Bing able to pass in symbols solves the first problem, but not the second. I get that an enum class is not a typical solution in SC, I really do. But it seems to be a really elegant solution in this case.

**Comment by telephon** [comment]
that's a good idea. The method `asUGenInput` could then just return the number.

**Comment by snappizz** [comment]
#2603 (abandoned) has my attempt at giving them names, which I'll replicate here:

```
		doneActionNames = IdentityDictionary[
			\noop -> 0,
			\pauseSelf -> 1,
			\freeSelf -> 2,
			\free -> 2, // shorthand because it's so common
			\freeSelfAndPrecedingNode -> 3,
			\freeSelfAndFollowingNode -> 4,
			\freeSelfAndNodesInPrecedingGroup -> 5,
			\freeSelfAndNodesInFollowingGroup -> 6,
			\freeSelfAndPrecedingNodesInGroup -> 7,
			\freeSelfAndFollowingNodesInGroup -> 8,
			\freeSelfAndPausePrecedingNode -> 9,
			\freeSelfAndPauseFollowingNode -> 10,
			\freeSelfAndDeepFreePrecedingGroup -> 11,
			\freeSelfAndDeepFreeFollowingGroup -> 12,
			\freeAllNodesInGroup -> 13,
			\freeGroup -> 14
		];
```

**Comment by snappizz** [comment]
Now that I've actually attempted to implement the symbols proposal, I realize it's not such a good idea. I agree with Brian now.

**Comment by jamshark70** [comment]
Sure, there are trade-offs here:

DoneAction class:

- Pros: Clear; simple; doesn't require changes elsewhere; integrates easily with existing auto-completion.

- Cons: No validation of other input types (which, admittedly, we don't currently have -- but if we are changing it, why not consider more general input validation? `doneAction: 300`). It will increase the size of the method lookup table. I tried to measure this and it's just a fraction of a percentage point, but users on embedded systems might want to keep memory use as small as possible. So it might be good to make it optional.

asDoneAction method:

- Pros: SC-idiomatic. Serves as a hook to validate any input type.

- Cons: UGen authors need to be aware of the requirement. No IDE auto-completion support (but, I think it's an unpleasant limitation of the IDE that completion is limited to class and method names -- it would be very useful to allow the language to add arbitrary symbols for completion, and maybe this feature would be just the motivation to do it).

I don't have a strong feeling about it either way.

**Comment by vivid-synth** [comment]
- I love the idea of not requiring users to know a bunch of magic numbers, but to achieve that we need to be able to refer to the class or symbols in the docs. For example, it's clear we'd want the second or third option in the PlayBuf helpfile:

     - > ar (numChannels, ... doneAction: 0)

    - > ar (numChannels, ... doneAction: 'doNothing')

    - > ar (numChannels, ... doneAction: Done.doNothing)

  - ```.doNothing``` feels long for something simple - can we think of a shorter alias? E.g. ```.noop``` or ```.none``` ?

**Comment by brianlheim** [comment]
>we need to be able to refer to the class or symbols in the docs

Absolutely, I've added that as a todo under the PR I opened: #2616 . We should move discussion there.

> .doNothing feels long for something simple - can we think of a shorter alias? E.g. .noop or .none

Definitely, although see #2614.

**Comment by jamshark70** [comment]
> I love the idea of not requiring users to know a bunch of magic numbers

FWIW I don't remember ever seeing anybody use a done action other than 0 or 2. *Maybe* 1. Even with symbolic names, most of them are useful only for cases that are not especially convenient to get oneself into in the first place. I don't think the new feature is going to change that.

**Comment by brianlheim** [comment]
Closed by #2616
